KafkaBasedCommandSender might not receive command response if response topic didn't exist before

The `CommandAndControlAmqpIT#testSendCommandFailsWhenNoCredit` test sometimes fails:
````
[ERROR] testSendCommandFailsWhenNoCredit{AmqpCommandEndpointConfiguration, VertxTestContext}[2]  Time elapsed: 7.781 s  <<< FAILURE!
java.lang.AssertionError: 
java.lang.AssertionError: 
Expecting:
  "org.eclipse.hono.client.SendMessageTimeoutException: sending command timed out after 4600ms
	at org.eclipse.hono.tests.IntegrationTestSupport.lambda$sendCommand$21(IntegrationTestSupport.java:1133)
	at io.vertx.core.impl.VertxImpl$InternalTimerHandler.handle(VertxImpl.java:942)
[...]
"
not to be an instance of: org.eclipse.hono.client.SendMessageTimeoutException
	at org.eclipse.hono.tests.amqp.CommandAndControlAmqpIT.lambda$testSendCommandFailsWhenNoCredit$69(CommandAndControlAmqpIT.java:763)
	at org.eclipse.hono.tests.amqp.CommandAndControlAmqpIT.lambda$testSendCommandFailsWhenNoCredit$70(CommandAndControlAmqpIT.java:758)
````

Log output:
````
2021-06-18T14:18:47.9336314Z 14:18:47.932 [main] INFO  o.e.h.t.amqp.CommandAndControlAmqpIT - running testSendCommandFailsWhenNoCredit(AmqpCommandEndpointConfiguration, VertxTestContext) [2]; parameters: subscribe as: GATEWAY_FOR_ALL_DEVICES, southbound endpoint: command, northbound endpoint: command
2021-06-18T14:18:47.9338917Z 14:18:47.932 [main] INFO  o.e.h.tests.IntegrationTestSupport - Configured to connect to Kafka on localhost:34277
2021-06-18T14:18:47.9345257Z 14:18:47.933 [main] INFO  o.e.h.tests.IntegrationTestSupport - Configured to connect to Kafka on localhost:34277
2021-06-18T14:18:47.9401644Z 14:18:47.939 [main] INFO  o.e.h.a.SpringBasedHonoPasswordEncoder - using BCrypt [strength: 4] with PRNG [NativePRNGNonBlocking] for encoding clear text passwords
2021-06-18T14:18:47.9482431Z 14:18:47.947 [main] INFO  o.e.h.a.SpringBasedHonoPasswordEncoder - using BCrypt [strength: 4] with PRNG [NativePRNGNonBlocking] for encoding clear text passwords
2021-06-18T14:18:47.9744999Z 14:18:47.973 [vert.x-kafka-consumer-thread-167] WARN  o.apache.kafka.clients.NetworkClient - [Consumer clientId=event-0ba547d8-fc4a-4948-b1f3-a84a83665433, groupId=its-9709fdaf-ef8c-4a1b-8425-12ffcb42ee14] Error while fetching metadata with correlation id 2 : {hono.event.a1bf5a70-2fc9-4c5a-a07f-5e470cd7985b=LEADER_NOT_AVAILABLE}
2021-06-18T14:18:48.1404370Z 14:18:48.139 [vert.x-eventloop-thread-2] INFO  o.e.h.t.amqp.CommandAndControlAmqpIT - AMQPS connection to adapter [localhost:49176] established
2021-06-18T14:18:51.0585766Z 14:18:51.057 [vert.x-eventloop-thread-1] WARN  o.e.h.a.c.k.i.KafkaBasedCommandSender - [auto.offset.reset] value is set to other than [latest]. It will be ignored and internally set to [latest]

2021-06-18T14:18:51.0783260Z 14:18:51.074 [vert.x-kafka-consumer-thread-168] WARN  o.apache.kafka.clients.NetworkClient - [Consumer clientId=command_response-ee0779bf-4c18-4f68-b53e-768c3f69ae60, groupId=a1bf5a70-2fc9-4c5a-a07f-5e470cd7985b-750e0d48-7f1b-4279-964d-c4fb9cad092a] Error while fetching metadata with correlation id 2 : {hono.command_response.a1bf5a70-2fc9-4c5a-a07f-5e470cd7985b=LEADER_NOT_AVAILABLE}

2021-06-18T14:18:51.8736854Z MQTT14:18:51.872 [kafka-producer-network-thread | Hono MQTT Adapter-event-172e7629-c175-4a8a-8e74-19096c42c3df] WARN  o.apache.kafka.clients.NetworkClient - [Producer clientId=Hono MQTT Adapter-event-172e7629-c175-4a8a-8e74-19096c42c3df] Error while fetching metadata with correlation id 1308 : {hono.event.ff745dc0-f404-4b97-a28a-7068366ec841=LEADER_NOT_AVAILABLE, hono.event.f6e0747c-0536-44f5-8767-787da6102d16=LEADER_NOT_AVAILABLE}

2021-06-18T14:18:55.6700723Z 14:18:55.666 [main] INFO  o.e.h.tests.IntegrationTestSupport - Configured to connect to Kafka on localhost:34277
2021-06-18T14:18:55.6817039Z 14:18:55.680 [vert.x-eventloop-thread-2] INFO  o.e.h.t.amqp.CommandAndControlAmqpIT - connection to AMQP adapter closed
````

--- a/clients/kafka-common/src/main/java/org/eclipse/hono/client/kafka/consumer/HonoKafkaConsumer.java
+++ b/clients/kafka-common/src/main/java/org/eclipse/hono/client/kafka/consumer/HonoKafkaConsumer.java
@@ -14,6 +14,7 @@ package org.eclipse.hono.client.kafka.consumer;
 
 import java.lang.reflect.Field;
 import java.net.HttpURLConnection;
+import java.time.Duration;
 import java.util.Collection;
 import java.util.HashSet;
 import java.util.List;
@@ -74,12 +75,14 @@ public class HonoKafkaConsumer implements Lifecycle {
      */
     private static final String DEFAULT_MAX_POLL_INTERNAL_MS = Long.toString(TimeUnit.SECONDS.toMillis(20));
 
+    private static final Duration POLL_TIMEOUT = Duration.ofSeconds(1);
+
     protected final Logger log = LoggerFactory.getLogger(getClass());
     protected final Vertx vertx;
     protected final AtomicBoolean stopCalled = new AtomicBoolean();
 
     private final Map<String, String> consumerConfig;
-    private final AtomicReference<Promise<Void>> subscribeDonePromiseRef = new AtomicReference<>();
+    private final AtomicReference<Promise<Set<TopicPartition>>> subscribeDonePromiseRef = new AtomicReference<>();
     private final Handler<KafkaConsumerRecord<String, Buffer>> recordHandler;
     private final Set<String> topics;
     private final Pattern topicPattern;
@@ -266,39 +269,77 @@ public class HonoKafkaConsumer implements Lifecycle {
                     .map(supplier -> KafkaConsumer.create(vertx, supplier.get()))
                     .orElseGet(() -> KafkaConsumer.create(vertx, consumerConfig, String.class, Buffer.class));
             kafkaConsumer.handler(record -> {
-                if (respectTtl && KafkaRecordHelper.isTtlElapsed(record.headers())) {
-                    onRecordHandlerSkippedForExpiredRecord(record);
-                } else {
-                    try {
-                        recordHandler.handle(record);
-                    } catch (final Exception e) {
-                        log.warn("error handling record [topic: {}, partition: {}, offset: {}, headers: {}]",
-                                record.topic(), record.partition(), record.offset(), record.headers(), e);
-                    }
+                if (!startPromise.future().isComplete()) {
+                    log.debug("postponing record handling until start() is completed [topic: {}, partition: {}, offset: {}]",
+                            record.topic(), record.partition(), record.offset());
                 }
+                startPromise.future().onSuccess(v2 -> {
+                    if (respectTtl && KafkaRecordHelper.isTtlElapsed(record.headers())) {
+                        onRecordHandlerSkippedForExpiredRecord(record);
+                    } else {
+                        try {
+                            recordHandler.handle(record);
+                        } catch (final Exception e) {
+                            log.warn("error handling record [topic: {}, partition: {}, offset: {}, headers: {}]",
+                                    record.topic(), record.partition(), record.offset(), record.headers(), e);
+                        }
+                    }
+                });
             });
             kafkaConsumer.exceptionHandler(error -> log.error("consumer error occurred", error));
             installRebalanceListeners();
             // subscribe and wait for rebalance to make sure that when start() completes,
             // the consumer is actually ready to receive records already
+            kafkaConsumer.asStream().pollTimeout(Duration.ofMillis(10)); // let polls finish quickly until start() is completed
             subscribeAndWaitForRebalance()
-                    .map(ok -> {
-                        if (topicPattern != null) {
-                            if (subscribedTopicPatternTopics.size() <= 5) {
-                                log.debug("subscribed to topic pattern [{}], matching topics: {}", topicPattern,
-                                        subscribedTopicPatternTopics);
-                            } else {
-                                log.debug("subscribed to topic pattern [{}], matching {} topics", topicPattern,
-                                        subscribedTopicPatternTopics.size());
-                            }
-                        } else {
-                            log.debug("subscribed to topics {}", topics);
+                    .compose(assignedPartitions -> {
+                        if (!assignedPartitions.isEmpty()
+                                && "latest".equals(consumerConfig.get(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG))) {
+                            // Wait for offset positions to be reset
+                            // (otherwise records published between start() completion and the offset-reset might get skipped)
+                            // This is done by having KafkaConsumer.updateFetchPositions() be called.
+                            return updateFetchPositions(assignedPartitions);
                         }
-                        return (Void) null;
+                        return Future.succeededFuture((Void) null);
+                    })
+                    .onSuccess(v2 -> {
+                        kafkaConsumer.asStream().pollTimeout(POLL_TIMEOUT);
+                        logSubscribedTopicsOnStartComplete();
                     })
                     .onComplete(startPromise.future());
         });
         return startPromise.future();
+    }
+
+    private Future<Void> updateFetchPositions(final Set<TopicPartition> assignedPartitions) {
+        final Promise<Void> resultPromise = Promise.promise();
+        runOnKafkaWorkerThread(v -> {
+            final Consumer<String, Buffer> wrappedConsumer = kafkaConsumer.asStream().unwrap();
+            try {
+                log.trace("fetching positions for {} assigned partitions...", assignedPartitions.size());
+                assignedPartitions.forEach(partition -> wrappedConsumer.position(Helper.to(partition)));
+                log.trace("done fetching positions for {} assigned partitions...", assignedPartitions.size());
+            } catch (final Exception e) {
+                log.error("error fetching positions for {} assigned partitions", assignedPartitions.size(), e);
+            } finally {
+                runOnContext(v2 -> resultPromise.complete());
+            }
+        });
+        return resultPromise.future();
+    }
+
+    private void logSubscribedTopicsOnStartComplete() {
+        if (topicPattern != null) {
+            if (subscribedTopicPatternTopics.size() <= 5) {
+                log.debug("consumer started, subscribed to topic pattern [{}], matching topics: {}", topicPattern,
+                        subscribedTopicPatternTopics);
+            } else {
+                log.debug("consumer started, subscribed to topic pattern [{}], matching {} topics", topicPattern,
+                        subscribedTopicPatternTopics.size());
+            }
+        } else {
+            log.debug("consumer started, subscribed to topics {}", topics);
+        }
     }
 
     /**
@@ -358,16 +399,31 @@ public class HonoKafkaConsumer implements Lifecycle {
         });
     }
 
-    private Future<Void> subscribeAndWaitForRebalance() {
+    private Future<Set<TopicPartition>> subscribeAndWaitForRebalance() {
         if (stopCalled.get()) {
             return Future.failedFuture(new ServerErrorException(HttpURLConnection.HTTP_UNAVAILABLE, "already stopped"));
         }
-        final Promise<Void> subscribeDonePromise = subscribeDonePromiseRef
+        final Promise<Set<TopicPartition>> subscribeDonePromise = subscribeDonePromiseRef
                 .updateAndGet(promise -> promise == null ? Promise.promise() : promise);
         final Promise<Void> subscriptionPromise = Promise.promise();
         if (topicPattern != null) {
             kafkaConsumer.subscribe(topicPattern, subscriptionPromise);
         } else {
+            // Trigger retrieval of metadata for each of the subscription topics if not already available locally;
+            // this will also trigger topic auto-creation if the topic doesn't exist yet.
+            // Doing so before the "subscribe" invocation shall ensure that these partitions are considered for
+            // partition assignment.
+            topics.forEach(topic -> {
+                final Promise<List<PartitionInfo>> partitionsForFuture = Promise.promise();
+                partitionsForFuture.future()
+                        .onSuccess(partitions -> {
+                            if (partitions.isEmpty()) {
+                                log.info("subscription topic doesn't exist and didn't get auto-created: {}", topic);
+                            }
+                        });
+                HonoKafkaConsumerHelper.partitionsFor(kafkaConsumer, topic, partitionsForFuture);
+            });
+
             kafkaConsumer.subscribe(topics, subscriptionPromise);
         }
         // kafkaConsumerWorker has to be retrieved after the first "subscribe" invocation
@@ -381,7 +437,8 @@ public class HonoKafkaConsumer implements Lifecycle {
                 subscribeDonePromise.tryFail(new ServerErrorException(HttpURLConnection.HTTP_UNAVAILABLE, errorMsg));
             }
         });
-        return CompositeFuture.all(subscriptionPromise.future(), subscribeDonePromise.future()).mapEmpty();
+        return CompositeFuture.all(subscriptionPromise.future(), subscribeDonePromise.future())
+                .map(r -> subscribeDonePromise.future().result());
     }
 
     /**
@@ -413,12 +470,12 @@ public class HonoKafkaConsumer implements Lifecycle {
             subscriptionResultPromise.future()
                     .onSuccess(result -> subscribedTopicPatternTopics = new HashSet<>(result))
                     .onFailure(thr -> log.info("failed to get subscription", thr))
-                    .map((Void) null)
+                    .map(partitionsSet)
                     .onComplete(ar -> Optional.ofNullable(subscribeDonePromiseRef.getAndSet(null))
                             .ifPresent(promise -> tryCompletePromise(promise, ar)));
         } else {
             Optional.ofNullable(subscribeDonePromiseRef.getAndSet(null))
-                    .ifPresent(Promise::tryComplete);
+                    .ifPresent(promise -> promise.tryComplete(partitionsSet));
         }
         if (onPartitionsAssignedHandler != null) {
             onPartitionsAssignedHandler.handle(partitionsSet);
@@ -595,12 +652,13 @@ public class HonoKafkaConsumer implements Lifecycle {
                     // therefore enforce a refresh here by again subscribing to the topic pattern
                     log.debug("ensureTopicIsAmongSubscribedTopics: verified topic existence, wait for subscription update and rebalance [{}]", topic);
                     return subscribeAndWaitForRebalance()
+                            .map((Void) null)
                             .compose(v -> {
                                 if (!subscribedTopicPatternTopics.contains(topic)) {
                                     // first metadata refresh could have failed with a LEADER_NOT_AVAILABLE error for the new topic;
                                     // seems to happen when some other topics have just been deleted for example
                                     log.debug("ensureTopicIsAmongSubscribedTopics: subscription not updated with topic after rebalance; try again [topic: {}]", topic);
-                                    return subscribeAndWaitForRebalance();
+                                    return subscribeAndWaitForRebalance().mapEmpty();
                                 }
                                 return Future.succeededFuture(v);
                             })

--- a/tests/src/test/java/org/eclipse/hono/tests/client/HonoKafkaConsumerIT.java
+++ b/tests/src/test/java/org/eclipse/hono/tests/client/HonoKafkaConsumerIT.java
@@ -1,0 +1,342 @@
+/*******************************************************************************
+ * Copyright (c) 2021 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+package org.eclipse.hono.tests.client;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.UUID;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import org.apache.kafka.clients.consumer.ConsumerConfig;
+import org.eclipse.hono.client.kafka.consumer.HonoKafkaConsumer;
+import org.eclipse.hono.tests.AssumeMessagingSystem;
+import org.eclipse.hono.tests.IntegrationTestSupport;
+import org.eclipse.hono.util.MessagingType;
+import org.junit.jupiter.api.AfterAll;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.vertx.core.CompositeFuture;
+import io.vertx.core.Future;
+import io.vertx.core.Handler;
+import io.vertx.core.Promise;
+import io.vertx.core.Vertx;
+import io.vertx.core.buffer.Buffer;
+import io.vertx.junit5.Timeout;
+import io.vertx.junit5.VertxExtension;
+import io.vertx.junit5.VertxTestContext;
+import io.vertx.kafka.admin.KafkaAdminClient;
+import io.vertx.kafka.admin.NewTopic;
+import io.vertx.kafka.client.consumer.KafkaConsumerRecord;
+import io.vertx.kafka.client.producer.KafkaProducer;
+import io.vertx.kafka.client.producer.KafkaProducerRecord;
+import io.vertx.kafka.client.producer.RecordMetadata;
+
+/**
+ * Test cases verifying the behavior of {@link HonoKafkaConsumer}.
+ * <p>
+ * To run this on a specific Kafka cluster instance, set the
+ * {@value IntegrationTestSupport#PROPERTY_DOWNSTREAM_BOOTSTRAP_SERVERS} system property,
+ * e.g. <code>-Ddownstream.bootstrap.servers="PLAINTEXT://localhost:9092"</code>.
+ */
+@ExtendWith(VertxExtension.class)
+@AssumeMessagingSystem(type = MessagingType.kafka)
+public class HonoKafkaConsumerIT {
+
+    private static final Logger LOG = LoggerFactory.getLogger(HonoKafkaConsumerIT.class);
+
+    private static final short REPLICATION_FACTOR = 1;
+
+    private static Vertx vertx;
+    private static KafkaAdminClient adminClient;
+    private static KafkaProducer<String, Buffer> kafkaProducer;
+    private static List<String> topicsToDeleteAfterTests = new ArrayList<>();
+
+    private HonoKafkaConsumer kafkaConsumer;
+
+    /**
+     * Sets up fixture.
+     */
+    @BeforeAll
+    public static void init() {
+        vertx = Vertx.vertx();
+
+        final Map<String, String> adminClientConfig = IntegrationTestSupport.getKafkaAdminClientConfig()
+                .getAdminClientConfig("test");
+        adminClient = KafkaAdminClient.create(vertx, adminClientConfig);
+        final Map<String, String> producerConfig = IntegrationTestSupport.getKafkaProducerConfig()
+                .getProducerConfig("test");
+        kafkaProducer = KafkaProducer.create(vertx, producerConfig);
+    }
+
+    /**
+     * Cleans up fixture.
+     *
+     * @param ctx The vert.x test context.
+     */
+    @AfterAll
+    public static void shutDown(final VertxTestContext ctx) {
+        final Promise<Void> producerClosePromise = Promise.promise();
+        kafkaProducer.close(producerClosePromise);
+
+        final Promise<Void> topicsDeletedPromise = Promise.promise();
+        adminClient.deleteTopics(topicsToDeleteAfterTests, topicsDeletedPromise);
+        topicsDeletedPromise.future()
+                .recover(thr -> {
+                    LOG.info("error deleting topics", thr);
+                    return Future.succeededFuture();
+                })
+                .compose(ar -> producerClosePromise.future())
+                .onComplete(ar -> {
+                    topicsToDeleteAfterTests.clear();
+                    topicsToDeleteAfterTests = null;
+                    adminClient.close();
+                    adminClient = null;
+                    kafkaProducer = null;
+                    vertx.close();
+                    vertx = null;
+                })
+                .onComplete(ctx.completing());
+    }
+
+    /**
+     * Closes a Kafka consumer created during the test.
+     *
+     * @param ctx The vert.x test context.
+     */
+    @AfterEach
+    void closeConsumer(final VertxTestContext ctx) {
+        if (kafkaConsumer != null) {
+            kafkaConsumer.stop().onComplete(ctx.completing());
+        }
+    }
+
+    /**
+     * Verifies that a HonoKafkaConsumer configured with "latest" as offset reset strategy only receives
+     * records published after the consumer <em>start()</em> method has completed.
+     *
+     * @param ctx The vert.x test context.
+     * @throws InterruptedException if test execution gets interrupted.
+     */
+    @Test
+    @Timeout(value = 10, timeUnit = TimeUnit.SECONDS)
+    public void testConsumerReadsLatestRecordsPublishedAfterStart(final VertxTestContext ctx) throws InterruptedException {
+        final int numTopics = 2;
+        final int numPartitions = 5;
+        final int numTestRecordsPerTopic = 20;
+
+        final Set<String> topics = IntStream.range(0, numTopics)
+                .mapToObj(i -> "test_" + i + "_" + UUID.randomUUID())
+                .collect(Collectors.toSet());
+
+        final VertxTestContext setup = new VertxTestContext();
+        createTopics(topics, numPartitions)
+                .compose(v -> publishRecords(numTestRecordsPerTopic, "key_", topics))
+                .onComplete(setup.completing());
+
+        assertThat(setup.awaitCompletion(IntegrationTestSupport.getTestSetupTimeout(), TimeUnit.SECONDS)).isTrue();
+        if (setup.failed()) {
+            ctx.failNow(setup.causeOfFailure());
+            return;
+        }
+        LOG.debug("topics created and test records published");
+
+        // prepare consumer
+        final Map<String, String> consumerConfig = IntegrationTestSupport.getKafkaConsumerConfig().getConsumerConfig("test");
+        consumerConfig.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "latest");
+
+        final AtomicReference<Promise<Void>> nextRecordReceivedPromiseRef = new AtomicReference<>();
+        final List<KafkaConsumerRecord<String, Buffer>> receivedRecords = new ArrayList<>();
+        final Handler<KafkaConsumerRecord<String, Buffer>> recordHandler = record -> {
+            receivedRecords.add(record);
+            Optional.ofNullable(nextRecordReceivedPromiseRef.get())
+                    .ifPresent(Promise::complete);
+        };
+        kafkaConsumer = new HonoKafkaConsumer(vertx, topics, recordHandler, consumerConfig);
+        // start consumer
+        kafkaConsumer.start().onComplete(ctx.succeeding(v -> {
+            ctx.verify(() -> {
+                assertThat(receivedRecords.size()).isEqualTo(0);
+            });
+            final Promise<Void> nextRecordReceivedPromise = Promise.promise();
+            nextRecordReceivedPromiseRef.set(nextRecordReceivedPromise);
+
+            LOG.debug("consumer started, publish record to be received by the consumer");
+            final String recordKey = "addedAfterStartKey";
+            publish(topics.iterator().next(), recordKey, Buffer.buffer("testPayload"));
+
+            nextRecordReceivedPromise.future().onComplete(ar -> {
+                ctx.verify(() -> {
+                    assertThat(receivedRecords.size()).isEqualTo(1);
+                    assertThat(receivedRecords.get(0).key()).isEqualTo(recordKey);
+                });
+                ctx.completeNow();
+            });
+        }));
+    }
+
+    /**
+     * Verifies that a HonoKafkaConsumer configured with "earliest" as offset reset strategy receives all
+     * current records after the consumer <em>start()</em> method has completed.
+     *
+     * @param ctx The vert.x test context.
+     * @throws InterruptedException if test execution gets interrupted.
+     */
+    @Test
+    @Timeout(value = 10, timeUnit = TimeUnit.SECONDS)
+    public void testConsumerReadsAllRecordsAfterStart(final VertxTestContext ctx) throws InterruptedException {
+        final int numTopics = 2;
+        final int numPartitions = 5;
+        final int numTestRecordsPerTopic = 20;
+
+        final Set<String> topics = IntStream.range(0, numTopics)
+                .mapToObj(i -> "test_" + i + "_" + UUID.randomUUID())
+                .collect(Collectors.toSet());
+
+        final VertxTestContext setup = new VertxTestContext();
+        createTopics(topics, numPartitions)
+                .compose(v -> publishRecords(numTestRecordsPerTopic, "key_", topics))
+                .onComplete(setup.completing());
+
+        assertThat(setup.awaitCompletion(IntegrationTestSupport.getTestSetupTimeout(), TimeUnit.SECONDS)).isTrue();
+        if (setup.failed()) {
+            ctx.failNow(setup.causeOfFailure());
+            return;
+        }
+        LOG.debug("topics created and test records published");
+
+        // prepare consumer
+        final Map<String, String> consumerConfig = IntegrationTestSupport.getKafkaConsumerConfig().getConsumerConfig("test");
+        consumerConfig.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
+
+        final Promise<Void> allRecordsReceivedPromise = Promise.promise();
+        final List<KafkaConsumerRecord<String, Buffer>> receivedRecords = new ArrayList<>();
+        final int totalExpectedMessages = numTopics * numTestRecordsPerTopic;
+        final Handler<KafkaConsumerRecord<String, Buffer>> recordHandler = record -> {
+            receivedRecords.add(record);
+            if (receivedRecords.size() == totalExpectedMessages) {
+                allRecordsReceivedPromise.complete();
+            }
+        };
+        kafkaConsumer = new HonoKafkaConsumer(vertx, topics, recordHandler, consumerConfig);
+        // start consumer
+        kafkaConsumer.start().onComplete(ctx.succeeding(v -> {
+            ctx.verify(() -> {
+                assertThat(receivedRecords.size()).isEqualTo(0);
+            });
+            allRecordsReceivedPromise.future().onComplete(ar -> {
+                ctx.verify(() -> {
+                    assertThat(receivedRecords.size()).isEqualTo(totalExpectedMessages);
+                });
+                ctx.completeNow();
+            });
+        }));
+    }
+
+    /**
+     * Verifies that a HonoKafkaConsumer that is using a not yet existing topic and that is configured with
+     * "latest" as offset reset strategy, only receives records on the auto-created topic published after the consumer
+     * <em>start()</em> method has completed.
+     *
+     * @param ctx The vert.x test context.
+     */
+    @Test
+    @Timeout(value = 10, timeUnit = TimeUnit.SECONDS)
+    public void testConsumerAutoCreatesTopicAndReadsLatestRecordsPublishedAfterStart(final VertxTestContext ctx) {
+
+        // prepare consumer
+        final Map<String, String> consumerConfig = IntegrationTestSupport.getKafkaConsumerConfig().getConsumerConfig("test");
+        consumerConfig.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "latest");
+
+        final AtomicReference<Promise<Void>> nextRecordReceivedPromiseRef = new AtomicReference<>();
+        final List<KafkaConsumerRecord<String, Buffer>> receivedRecords = new ArrayList<>();
+        final Handler<KafkaConsumerRecord<String, Buffer>> recordHandler = record -> {
+            receivedRecords.add(record);
+            Optional.ofNullable(nextRecordReceivedPromiseRef.get())
+                    .ifPresent(Promise::complete);
+        };
+        final String topic = "test_" + UUID.randomUUID();
+        topicsToDeleteAfterTests.add(topic);
+        kafkaConsumer = new HonoKafkaConsumer(vertx, Set.of(topic), recordHandler, consumerConfig);
+        // start consumer
+        kafkaConsumer.start().onComplete(ctx.succeeding(v -> {
+            ctx.verify(() -> {
+                assertThat(receivedRecords.size()).isEqualTo(0);
+            });
+            final Promise<Void> nextRecordReceivedPromise = Promise.promise();
+            nextRecordReceivedPromiseRef.set(nextRecordReceivedPromise);
+
+            LOG.debug("consumer started, publish record to be received by the consumer");
+            final String recordKey = "addedAfterStartKey";
+            publish(topic, recordKey, Buffer.buffer("testPayload"));
+
+            nextRecordReceivedPromise.future().onComplete(ar -> {
+                ctx.verify(() -> {
+                    assertThat(receivedRecords.size()).isEqualTo(1);
+                    assertThat(receivedRecords.get(0).key()).isEqualTo(recordKey);
+                });
+                ctx.completeNow();
+            });
+        }));
+    }
+
+    private static Future<Void> createTopics(final Collection<String> topicNames, final int numPartitions) {
+        topicsToDeleteAfterTests.addAll(topicNames);
+        final Promise<Void> resultPromise = Promise.promise();
+        final List<NewTopic> topics = topicNames.stream()
+                .map(t -> new NewTopic(t, numPartitions, REPLICATION_FACTOR))
+                .collect(Collectors.toList());
+        adminClient.createTopics(topics, resultPromise.future());
+        return resultPromise.future();
+    }
+
+    private Future<Void> publishRecords(final int numTestRecordsPerTopic, final String keyPrefix, final Set<String> topics) {
+        @SuppressWarnings("rawtypes")
+        final List<Future> resultFutures = new ArrayList<>();
+        topics.forEach(topic -> {
+            resultFutures.add(publishRecords(numTestRecordsPerTopic, keyPrefix, topic));
+        });
+        return CompositeFuture.all(resultFutures).map((Void) null);
+    }
+
+    private Future<Void> publishRecords(final int numRecords, final String keyPrefix, final String topic) {
+        @SuppressWarnings("rawtypes")
+        final List<Future> resultFutures = new ArrayList<>();
+        IntStream.range(0, numRecords).forEach(i -> {
+            resultFutures.add(publish(topic, keyPrefix + i, Buffer.buffer("testPayload")).mapEmpty());
+        });
+        return CompositeFuture.all(resultFutures).map((Void) null);
+    }
+
+    private static Future<RecordMetadata> publish(final String topic, final String recordKey, final Buffer recordPayload) {
+        final Promise<RecordMetadata> resultPromise = Promise.promise();
+        final KafkaProducerRecord<String, Buffer> record = KafkaProducerRecord.create(topic, recordKey, recordPayload);
+        kafkaProducer.send(record, resultPromise);
+        return resultPromise.future();
+    }
+
+}
+

